src/cryptonote_core/miner.cpp

```diff
@@ -408,5 +528,313 @@ namespace cryptonote
+  {
+    idle_time =
+      cpu_data_end.overall_cpu_times[CPU_DATA_OVERALL_IDLE_INDEX]
+      - cpu_data_start.overall_cpu_times[CPU_DATA_OVERALL_IDLE_INDEX];
+  }
+  //-----------------------------------------------------------------------------------------------------
+  void miner::get_cpu_active_time(const cpu_data& cpu_data_start, const cpu_data& cpu_data_end, uint64_t& active_time)
+  {
+    active_time =
+      (cpu_data_end.overall_cpu_times[CPU_DATA_OVERALL_USER_INDEX]
+      + cpu_data_end.overall_cpu_times[CPU_DATA_OVERALL_NICE_INDEX]
+      + cpu_data_end.overall_cpu_times[CPU_DATA_OVERALL_SYSTEM_INDEX])
+      -
+      (cpu_data_start.overall_cpu_times[CPU_DATA_OVERALL_USER_INDEX]
+      + cpu_data_start.overall_cpu_times[CPU_DATA_OVERALL_NICE_INDEX]
+      + cpu_data_start.overall_cpu_times[CPU_DATA_OVERALL_SYSTEM_INDEX]);
+  }
+  //-----------------------------------------------------------------------------------------------------
+  void miner::get_cpu_total_time(const cpu_data& cpu_data_start, const cpu_data& cpu_data_end, uint64_t& total_time)
+  {
+    uint64_t idle_time = 0, active_time = 0;
+    get_cpu_idle_time(cpu_data_start, cpu_data_end, idle_time);
+    get_cpu_active_time(cpu_data_start, cpu_data_end, active_time);
+    total_time = idle_time + active_time;
+  }
+  //-----------------------------------------------------------------------------------------------------
+  void miner::get_cpu_idle_percentage(const cpu_data& cpu_data_start, const cpu_data& cpu_data_end, uint8_t& idle_percentage)
+  {
+    uint64_t idle_time = 0, total_time = 0;
+    get_cpu_idle_time(cpu_data_start, cpu_data_end, idle_time);
+    get_cpu_total_time(cpu_data_start, cpu_data_end, total_time);
+    idle_percentage = (100.f * idle_time / total_time);
+  }
+  //-----------------------------------------------------------------------------------------------------
+  void miner::get_cpu_miner_percentage(const cpu_data& cpu_data_start, const cpu_data& cpu_data_end, uint8_t& miner_percentage)
+  {
+    uint64_t total_time = 0, miner_time = 0;
+    get_cpu_total_time(cpu_data_start, cpu_data_end, total_time);
+    miner_time = cpu_data_end.process_cpu_time - cpu_data_start.process_cpu_time;
+    miner_percentage = (100.f * miner_time / total_time);
+  }
+  //-----------------------------------------------------------------------------------------------------
+  bool miner::adjust_miner_extra_sleep(const cpu_data& cpu_data_start, const cpu_data& cpu_data_end, uint64_t& miner_extra_sleep)
+  {
+      if(cpu_data_start.update_time == 0 || cpu_data_end.update_time == 0)
+      {
+        if(cpu_data_start.update_time == 0 || cpu_data_end.update_time == 0)
+          LOG_PRINT_RED_L0("cpu_data_start is uninitialized");
+        if(cpu_data_start.update_time == 0 || cpu_data_end.update_time == 0)
+          LOG_PRINT_RED_L0("cpu_data_start is uninitialized");
```

Comment:
I don't follow this section - were other variables supposed to be checked? This looks to be checking the same 2 variables 3 times.